// oclif/lib/generators/hook.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("node:path");
const Generator = require("yeoman-generator");
const util_1 = require("../util");
const { version } = require('../../package.json');
class Hook extends Generator {
    options;
    pjson;
    constructor(args, options) {
        super(args, options);
        this.options = options;
    }
    async prompting() {
        this.pjson = this.fs.readJSON('package.json');
        this.pjson.oclif = this.pjson.oclif || {};
        if (!this.pjson)
            throw new Error('not in a project directory');
        this.log(`Adding a ${this.options.event} hook to ${this.pjson.name}! Version: ${version}`);
    }
    writing() {
        this.sourceRoot(path.join(__dirname, '../../templates'));
        this.fs.copyTpl(this.templatePath('src/hook.ts.ejs'), this.destinationPath(`src/hooks/${this.options.event}/${this.options.name}.ts`), this);
        if (this.hasMocha()) {
            this.fs.copyTpl(this.templatePath('test/hook.test.ts.ejs'), this.destinationPath(`test/hooks/${this.options.event}/${this.options.name}.test.ts`), this);
        }
        this.pjson.oclif = this.pjson.oclif || {};
        this.pjson.oclif.hooks = this.pjson.oclif.hooks || {};
        const hooks = this.pjson.oclif.hooks;
        const p = `./dist/hooks/${this.options.event}/${this.options.name}`;
        if (hooks[this.options.event]) {
            hooks[this.options.event] = (0, util_1.castArray)(hooks[this.options.event]);
            hooks[this.options.event] = [...hooks[this.options.event], p];
        }
        else {
            this.pjson.oclif.hooks[this.options.event] = p;
        }
        this.fs.writeJSON(this.destinationPath('./package.json'), this.pjson);
    }
    hasMocha() {
        return Boolean(this.pjson.devDependencies?.mocha);
    }
}
exports.default = Hook;
